Correcting the derivative control term.

ffarha opened the pull request:
<!--
If this PR is associated with an issue be sure to reference it
by including "refs #<issue>" or "closes #<issue>" (e.g., #closes #1234).
Issue #28217
If this PR implements an enhancement that does not have an existing issue, please add the following:
-->

## Reason
<!--Why do you need this feature or what is the enhancement?-->
C_{n} = C_{n-1} + K_{integral} \int_0^{t} pp(s) - target mathrm{d}s + K_{proportional} (pp(t) - target) + K_{derivative} \dfrac{pp(t) - target}{dt}
In the derivatve term, it should be using the derivative of the error (i.e., d(pp - target)/dt and not (pp - target) / delta_t), if only because the correction term would otherwise blow up as delta_t goes to 0.
## Design
<!--A concise description (design) of the enhancement.-->
The derivative calculation formula is corrected to include d(pp - target) instead of (pp - target).
## Impact
<!--Will the enhancement change existing APIs or add something new?-->
Change existing APIs.

<!--
If this PR implements a bug fix, please create an issue for the bug and reference the issue.
-->

ffarha committed:
Correcting the derivative control term, issue#28217

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/2351982/) on 883022b wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/28235/site)

This comment will be updated on new commits.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/2352142/) on 0627d6f wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>9a7e0e</th>
      <th colspan=4>#28235 0627d6</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.07%</td>
      <td>85.07%</td>
      <td>-</td>
      <td>100.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>104699</td>
      <td>104699</td>
      <td>-</td>
      <td>1</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>18375</td>
      <td>18375</td>
      <td>-</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/28235/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/28235/coverage/framework)


## Modules coverage

Coverage did not change

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/28235/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/28235/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/28235/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/28235/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/28235/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/28235/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/28235/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/28235/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/28235/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/28235/coverage/geochemistry)
- [`heat_transfer`](https://mooseframework.inl.gov/docs/PRs/28235/coverage/heat_transfer)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/28235/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/28235/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/28235/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/28235/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/28235/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/28235/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/28235/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/28235/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/28235/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/28235/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/28235/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/28235/coverage/scalar_transport)
- [`solid_mechanics`](https://mooseframework.inl.gov/docs/PRs/28235/coverage/solid_mechanics)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/28235/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/28235/coverage/stochastic_tools)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/28235/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/28235/coverage/xfem)

</details>



This comment will be updated on new commits.

GiudGiud commented on framework/src/controls/PIDTransientControl.C:
```diff
@@ -145,7 +145,7 @@ PIDTransientControl::execute()
     value += _Kint * _integral + _Kpro * delta;
     if (_dt > 0)
-      value += _Kder * delta / _dt;
+      value += _Kder * (delta - _old_delta) / _dt;
```
old delta is reset to 0 when the time step fails (see above)
you ll want to add additional logic to prevent using this 0

ffarha commented:
@GiudGiud tests that are not from PIDTransient control are failing in debug, does not look related to this PR.

ffarha commented on framework/src/controls/PIDTransientControl.C:
```diff
@@ -145,7 +147,12 @@ PIDTransientControl::execute()
     if (_dt > 0)
-      value += _Kder * delta / _dt;
+    {
+      if (_reset_every_timestep)
+        value += _Kder * (delta - _old_delta_unaffected) / _dt;
+      else
```
If the _reset_every_timestep is true, it is changing the _old_delta to 0 which cannot be used for the derivative term anymore. That's why in this case I have introduced another term _old_delta_affected. The _old_delta is saved in _old_delta_unaffected before resetting it to zero. The _old_delta_unaffected is then used for the calculation of the derivative term. 'If' statement because this only needs to be done if the _old_delta is reset.

GiudGiud replied:
> If the _reset_every_timestep is true, it is changing the _old_delta to 0 which cannot be used for the derivative term anymore.

on which line is this happening?

ffarha replied:
126, 127

GiudGiud replied:
it's not though
this
```
    if (_t_step != _t_step_old)
    {
      // Reset the error integral if PID is only used within each timestep
      if (_reset_every_timestep)
        _integral = 0;
      _integral_old = _integral;
      _value_old = value;
      _t_step_old = _t_step;
      _old_delta_unaffected = _old_delta;
      _old_delta = 0;
    }

```
is equivalent to this
```
    if (_t_step != _t_step_old)
    {
      // Reset the error integral if PID is only used within each timestep
      if (_reset_every_timestep)
      {
        _integral = 0;
      }
      _integral_old = _integral;
      _value_old = value;
      _t_step_old = _t_step;
      _old_delta_unaffected = _old_delta;
      _old_delta = 0;
    }
```
so _reset_every_timestep does not affect `_old_delta`

GiudGiud replied:
not sure
you could just use ` _old_delta_unaffected ` for that line all the time

I m not clear on whether ` _old_delta_unaffected ` is currenttly truly unaffected. I ll have to think more about it

ffarha replied:
alright I'm using _old_delta_unaffected all the time for derivative term now. please check.

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2356769/) on 38c48e5 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/28235/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/28235/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 89ffe9d1cf4836fb77572002e59c84f63b572365`

GiudGiud commented on framework/src/controls/PIDTransientControl.C:
```diff
@@ -170,5 +171,6 @@ PIDTransientControl::execute()
 
     // Keep track of the previous delta for integral windup control
     _old_delta = delta;
+    _old_delta_unaffected = delta;
```
this is being updated on failed solves. So it's not correct.

ffarha replied:
hi @GiudGiud, I think it is okay. The checks were failing because I had not updated a gold csv file before. Now all checks have passed.

GiudGiud replied:
I dont think it's ok. 
You can add a test using the Failing problem and a custom time stepper with FunctionDT

then you make the time stepper double one time step and the FailingProblem (a test object) fail that step and cut it back in half.
The output of the test should be the same as with a regular non-failing problem

ffarha replied:
hi @GiudGiud, I am leaving INL tomorrow morning and I'm afraid I won't be able to do it within this time. The derivate control is working with the adjustment so far with all the test inputs already implemented, it'd be great if you could consider having the improvement I've made.

GiudGiud replied:
I ended up investigating this and it wont work with a Failing problem.

however we don't have the solution just yet. I ll make it ready for this using a restartable attribute but it will only truly work when we recover attributes on failed time steps

GiudGiud commented:
the commit messages are not sufficient. "Addressing comments" does not tell me what it does

ffarha committed:
updating abs_max_rate_change.csv issue#28217

ffarha commented:
@sallustius

sallustius commented:
@GiudGiud, I'm keeping this PR alive. I can work on it in the upcoming weeks, unless @ffarha wants to complete it.

GiudGiud commented:
Yes that is a good idea, this is an important bug fix

GiudGiud committed:
Enable recover testing for PID Controller

GiudGiud committed:
Add support for recover mode to PIDTRansient
- logic for parameter values that dont get restarted
- delete the new 'unaffected' attribute, not useful
- make all the attributes restartable data
- dont ignore the bounds when using the maximum rate of change

GiudGiud committed:
Regold all PID transient test due to fixing the reset in delta_old that was not necessary

GiudGiud commented:
@joshuahansel  if you could please review

this PR ends up fixing the recover behavior as well, and correcting a few mistakes in the logic. Hence the regold. The additional recover testing gives me more confidence this is correct now

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/2754334/), step Docs: sync website on 386a978 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/28235/site)

This comment will be updated on new commits.

joshuahansel commented on framework/src/controls/PIDTransientControl.C:
```diff
@@ -109,14 +111,30 @@ PIDTransientControl::PIDTransientControl(const InputParameters & parameters)
+      // If just recovering, we must use the restartable value as the parameter value is
+      // not restarted
+      if (_app.isRecovering() && !_has_recovered)
+        _has_recovered = true;
```
This flag needs to be reset to false at some point to have the same behavior for subsequent recovers.

GiudGiud replied:
it's not restartable so it should reset itself on every recover

joshuahansel reviewed (changes requested):
Looks pretty good, thanks for the fix!

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/2754350/), step Generate coverage on 386a978 wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>3403d2</th>
      <th colspan=4>#28235 386a97</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.29%</td>
      <td>85.29%</td>
      <td>+0.00%</td>
      <td>100.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>109051</td>
      <td>109058</td>
      <td>+7</td>
      <td>23</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>18808</td>
      <td>18806</td>
      <td>-2</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/28235/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/28235/coverage/framework)


## Modules coverage

Mutiple base shas were found when retrieving head reports:
  a5ad50a5a7c3d77faf580763f48c8345b17fb06a
  3403d29f308c36cd2e06331e17c0ac33aafd6558

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/28235/coverage/framework)
- `chemical_reactions` (FAILED)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/28235/coverage/combined)
- `contact` (FAILED)
- `electromagnetics` (FAILED)
- `external_petsc_solver` (FAILED)
- `fluid_properties` (FAILED)
- `fsi` (FAILED)
- `functional_expansion_tools` (FAILED)
- `geochemistry` (FAILED)
- `heat_transfer` (FAILED)
- `level_set` (FAILED)
- `misc` (FAILED)
- `navier_stokes` (FAILED)
- `optimization` (FAILED)
- `peridynamics` (FAILED)
- `phase_field` (FAILED)
- `porous_flow` (FAILED)
- `ray_tracing` (FAILED)
- `rdg` (FAILED)
- `reactor` (FAILED)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/28235/coverage/richards)
- `scalar_transport` (FAILED)
- `solid_mechanics` (FAILED)
- `solid_properties` (FAILED)
- `stochastic_tools` (FAILED)
- `subchannel` (FAILED)
- `thermal_hydraulics` (FAILED)
- `xfem` (FAILED)

</details>



This comment will be updated on new commits.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/2754350/), step Verify coverage on 386a978 wanted to post the following:


The following coverage requirement(s) failed:
- Failed to generate `chemical_reactions` coverage rate (required: 92.0%)
- Failed to generate `contact` coverage rate (required: 87.0%)
- Failed to generate `electromagnetics` coverage rate (required: 94.0%)
- Failed to generate `external_petsc_solver` coverage rate (required: 84.0%)
- Failed to generate `fluid_properties` coverage rate (required: 84.0%)
- Failed to generate `fsi` coverage rate (required: 85.0%)
- Failed to generate `functional_expansion_tools` coverage rate (required: 81.0%)
- Failed to generate `geochemistry` coverage rate (required: 96.0%)
- Failed to generate `heat_transfer` coverage rate (required: 87.0%)
- Failed to generate `level_set` coverage rate (required: 85.0%)
- Failed to generate `misc` coverage rate (required: 32.0%)
- Failed to generate `navier_stokes` coverage rate (required: 77.0%)
- Failed to generate `optimization` coverage rate (required: 86.0%)
- Failed to generate `peridynamics` coverage rate (required: 77.0%)
- Failed to generate `phase_field` coverage rate (required: 85.0%)
- Failed to generate `porous_flow` coverage rate (required: 95.0%)
- Failed to generate `ray_tracing` coverage rate (required: 93.0%)
- Failed to generate `rdg` coverage rate (required: 63.0%)
- Failed to generate `reactor` coverage rate (required: 90.0%)
- Failed to generate `scalar_transport` coverage rate (required: 81.0%)
- Failed to generate `solid_mechanics` coverage rate (required: 84.0%)
- Failed to generate `solid_properties` coverage rate (required: 83.0%)
- Failed to generate `stochastic_tools` coverage rate (required: 88.0%)
- Failed to generate `subchannel` coverage rate (required: 87.0%)
- Failed to generate `thermal_hydraulics` coverage rate (required: 88.0%)
- Failed to generate `xfem` coverage rate (required: 80.0%)

GiudGiud commented:
openmpi failure unrelated

GiudGiud commented:
coverage failure is baseless slander

lindsayad commented:
This broke the sweeps in next

GiudGiud commented:
I'll fix it tomorrow. This is a control so it won't be too bad
